# **Data Preprocessing on Car dataset**

**Problem Statement:**
                     You are working for a youtube channel which does content about cars. So they have given you a dataset which holds informations about different types of cars. The data format is CSV file. They want you to create a new file with cars having weight less than 4000 and has origins from Asia and Europe

In [ ]:
from google.colab import drive
drive.mount('/content/drive',force_remount=True)

Mounted at /content/drive


Declaration of Necessary libraries

In [ ]:
import numpy as np 
import pandas as pd #used for handling data, cleaning and manipulation
import matplotlib.pyplot as plt #used for visualizations

Loading the Car dataset from google drive to Python file

In [ ]:
data=pd.read_csv('/content/drive/My Drive/DA_project/Cars_dataset.csv')

Taking a peek at how data looks

In [ ]:
data.head() #prints the first 5 data from the data set

Make           Model   Type Origin DriveTrain      MSRP   Invoice  \
0  Acura             MDX    SUV   Asia        All  $36,945   $33,337    
1  Acura  RSX Type S 2dr  Sedan   Asia      Front  $23,820   $21,761    
2  Acura         TSX 4dr  Sedan   Asia      Front  $26,990   $24,647    
3  Acura          TL 4dr  Sedan   Asia      Front  $33,195   $30,299    
4  Acura      3.5 RL 4dr  Sedan   Asia      Front  $43,755   $39,014    

   EngineSize  Cylinders  Horsepower  MPG_City  MPG_Highway  Weight  \
0         3.5        6.0       265.0      17.0         23.0  4451.0   
1         2.0        4.0       200.0      24.0         31.0  2778.0   
2         2.4        4.0       200.0      22.0         29.0  3230.0   
3         3.2        6.0       270.0      20.0         28.0  3575.0   
4         3.5        6.0       225.0      18.0         24.0  3880.0   

   Wheelbase  Length  
0      106.0   189.0  
1      101.0   172.0  
2      105.0   183.0  
3      108.0   186.0  
4      115.0   197.0

In [ ]:
data.shape #(Rows in data, Columns in data)

(432, 15)

In [ ]:
data.info() #The information about the data, coumns, its count and data type of each columns

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 432 entries, 0 to 431
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Make         428 non-null    object 
 1   Model        428 non-null    object 
 2   Type         428 non-null    object 
 3   Origin       428 non-null    object 
 4   DriveTrain   428 non-null    object 
 5   MSRP         428 non-null    object 
 6   Invoice      428 non-null    object 
 7   EngineSize   428 non-null    float64
 8   Cylinders    426 non-null    float64
 9   Horsepower   428 non-null    float64
 10  MPG_City     428 non-null    float64
 11  MPG_Highway  428 non-null    float64
 12  Weight       428 non-null    float64
 13  Wheelbase    428 non-null    float64
 14  Length       428 non-null    float64
dtypes: float64(8), object(7)
memory usage: 50.8+ KB


In [ ]:
data.isnull().sum() #looking for total count of null objects

Make           4
Model          4
Type           4
Origin         4
DriveTrain     4
MSRP           4
Invoice        4
EngineSize     4
Cylinders      6
Horsepower     4
MPG_City       4
MPG_Highway    4
Weight         4
Wheelbase      4
Length         4
dtype: int64

# **Data Cleaning and Preprocessing**

*** *The dataset consists of 13 columns. The dataset has both Qualitative data and Quantitative data. Four out of 13 columns - Make, Model, Type, Origin, DriveTrain are nominal data which is of qualitative type (i.e) we cannot perform any arithmetic operations on them but can use them as label or classifiers. So we can perform mean on the other columns to fill the Null data.* ***

Since there majorly 4 null datas, we can take the mean of the total column and put it to them. It will not be a big difference as 4 is negligible compared to 428 rows of data. This operation is shown below 

In [ ]:
data[['MSRP','Invoice','EngineSize','Cylinders','Horsepower','MPG_City','MPG_Highway','Weight','Wheelbase','Length']] = data[['MSRP','Invoice','EngineSize','Cylinders','Horsepower','MPG_City','MPG_Highway','Weight','Wheelbase','Length']].fillna(data[['MSRP','Invoice','EngineSize','Cylinders','Horsepower','MPG_City','MPG_Highway','Weight','Wheelbase','Length']].mean())

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  """Entry point for launching an IPython kernel.


In [ ]:
data.isnull().sum()

Make           4
Model          4
Type           4
Origin         4
DriveTrain     4
MSRP           4
Invoice        4
EngineSize     0
Cylinders      0
Horsepower     0
MPG_City       0
MPG_Highway    0
Weight         0
Wheelbase      0
Length         0
dtype: int64

From the above Null data count output we can see that the MSRP and Invoice columns still contains the null values even after replacing the mean values. By checking the dataset once again, those two columnns contains $ symbol which makes them a "String data type"(Something Like a word). So we need to remove the symble. The code for removing the symbol and filling null values with mean value is given below 

In [ ]:
data['MSRP'] = data['MSRP'].replace("[$,]", "", regex=True).astype(float)        #removes the $symbol in column MSRP
data['Invoice'] = data['Invoice'].replace("[$,]", "", regex=True).astype(float)  #removes the $symbol in column Invoice
data[['MSRP','Invoice']]= data[['MSRP','Invoice']].fillna(data[['MSRP','Invoice']].mean())

In [ ]:
data.isnull().sum()

Make           4
Model          4
Type           4
Origin         4
DriveTrain     4
MSRP           0
Invoice        0
EngineSize     0
Cylinders      0
Horsepower     0
MPG_City       0
MPG_Highway    0
Weight         0
Wheelbase      0
Length         0
dtype: int64

From the above output, the MSRP and Invoice columns are sorted. Now we need to work on the nominal data. Since 4 datas are negligible compared to 428, we fill the 4 null data with most occured data. This method works well when there is a negligible null data( like we have in our case) 

In [ ]:
clean_data = data.apply(lambda x: x.fillna(x.value_counts().index[0]))
clean_data

Make                    Model   Type  Origin DriveTrain      MSRP  \
0    Acura                      MDX    SUV    Asia        All  $36,945    
1    Acura           RSX Type S 2dr  Sedan    Asia      Front  $23,820    
2    Acura                  TSX 4dr  Sedan    Asia      Front  $26,990    
3    Acura                   TL 4dr  Sedan    Asia      Front  $33,195    
4    Acura               3.5 RL 4dr  Sedan    Asia      Front  $43,755    
..     ...                      ...    ...     ...        ...       ...   
427  Volvo  C70 LPT convertible 2dr  Sedan  Europe      Front  $40,565    
428  Volvo  C70 HPT convertible 2dr  Sedan  Europe      Front  $42,565    
429  Volvo               S80 T6 4dr  Sedan  Europe      Front  $45,210    
430  Volvo                      V40  Wagon  Europe      Front  $26,135    
431  Volvo                     XC70  Wagon  Europe        All  $35,145    

      Invoice  EngineSize  Cylinders  Horsepower  MPG_City  MPG_Highway  \
0    $33,337          3.5        6.0       265.0      17.0         23.0   
1    $21,761          2.0        4.0       200.0      24.0         31.0   
2    $24,647          2.4        4.0       200.0      22.0         29.0   
3    $30,299          3.2        6.0       270.0      20.0         28.0   
4    $39,014          3.5        6.0       225.0      18.0         24.0   
..        ...         ...        ...         ...       ...          ...   
427  $38,203          2.4        5.0       197.0      21.0         28.0   
428  $40,083          2.3        5.0       242.0      20.0         26.0   
429  $42,573          2.9        6.0       268.0      19.0         26.0   
430  $24,641          1.9        4.0       170.0      22.0         29.0   
431  $33,112          2.5        5.0       208.0      20.0         27.0   

     Weight  Wheelbase  Length  
0    4451.0      106.0   189.0  
1    2778.0      101.0   172.0  
2    3230.0      105.0   183.0  
3    3575.0      108.0   186.0  
4    3880.0      115.0   197.0  
..      ...        ...     ...  
427  3450.0      105.0   186.0  
428  3450.0      105.0   186.0  
429  3653.0      110.0   190.0  
430  2822.0      101.0   180.0  
431  3823.0      109.0   186.0  

[432 rows x 15 columns]

Checking for null value


In [ ]:
clean_data.isnull().sum()

Make           0
Model          0
Type           0
Origin         0
DriveTrain     0
MSRP           0
Invoice        0
EngineSize     0
Cylinders      0
Horsepower     0
MPG_City       0
MPG_Highway    0
Weight         0
Wheelbase      0
Length         0
dtype: int64

Analyzing different Car Makes

In [ ]:
clean_data['Make'].value_counts()

Toyota           32
Chevrolet        27
Mercedes-Benz    26
Ford             23
BMW              20
Audi             19
Nissan           17
Honda            17
Chrysler         15
Volkswagen       15
Mitsubishi       13
Dodge            13
Hyundai          12
Jaguar           12
Volvo            12
Kia              11
Mazda            11
Lexus            11
Pontiac          11
Subaru           11
Lincoln           9
Mercury           9
Buick             9
Saturn            8
Infiniti          8
GMC               8
Cadillac          8
Suzuki            8
Porsche           7
Saab              7
Acura             7
Oldsmobile        3
Jeep              3
Land Rover        3
MINI              2
Scion             2
Isuzu             2
Hummer            1
Name: Make, dtype: int64

Looking for cars from Asia and Europe

In [ ]:
origin_cars = clean_data[clean_data['Origin'].isin(['Asia','Europe'])] #this checks for the cars that are in asia and europe
origin_cars['Make'].value_counts()

Toyota           32
Mercedes-Benz    26
BMW              20
Audi             19
Honda            17
Nissan           17
Volkswagen       15
Mitsubishi       13
Volvo            12
Hyundai          12
Jaguar           12
Mazda            11
Kia              11
Lexus            11
Subaru           11
Suzuki            8
Infiniti          8
Saab              7
Acura             7
Porsche           7
Land Rover        3
Scion             2
Isuzu             2
MINI              2
Name: Make, dtype: int64

In [ ]:
origin_cars

Make                    Model   Type  Origin DriveTrain      MSRP  \
0    Acura                      MDX    SUV    Asia        All  $36,945    
1    Acura           RSX Type S 2dr  Sedan    Asia      Front  $23,820    
2    Acura                  TSX 4dr  Sedan    Asia      Front  $26,990    
3    Acura                   TL 4dr  Sedan    Asia      Front  $33,195    
4    Acura               3.5 RL 4dr  Sedan    Asia      Front  $43,755    
..     ...                      ...    ...     ...        ...       ...   
427  Volvo  C70 LPT convertible 2dr  Sedan  Europe      Front  $40,565    
428  Volvo  C70 HPT convertible 2dr  Sedan  Europe      Front  $42,565    
429  Volvo               S80 T6 4dr  Sedan  Europe      Front  $45,210    
430  Volvo                      V40  Wagon  Europe      Front  $26,135    
431  Volvo                     XC70  Wagon  Europe        All  $35,145    

      Invoice  EngineSize  Cylinders  Horsepower  MPG_City  MPG_Highway  \
0    $33,337          3.5        6.0       265.0      17.0         23.0   
1    $21,761          2.0        4.0       200.0      24.0         31.0   
2    $24,647          2.4        4.0       200.0      22.0         29.0   
3    $30,299          3.2        6.0       270.0      20.0         28.0   
4    $39,014          3.5        6.0       225.0      18.0         24.0   
..        ...         ...        ...         ...       ...          ...   
427  $38,203          2.4        5.0       197.0      21.0         28.0   
428  $40,083          2.3        5.0       242.0      20.0         26.0   
429  $42,573          2.9        6.0       268.0      19.0         26.0   
430  $24,641          1.9        4.0       170.0      22.0         29.0   
431  $33,112          2.5        5.0       208.0      20.0         27.0   

     Weight  Wheelbase  Length  
0    4451.0      106.0   189.0  
1    2778.0      101.0   172.0  
2    3230.0      105.0   183.0  
3    3575.0      108.0   186.0  
4    3880.0      115.0   197.0  
..      ...        ...     ...  
427  3450.0      105.0   186.0  
428  3450.0      105.0   186.0  
429  3653.0      110.0   190.0  
430  2822.0      101.0   180.0  
431  3823.0      109.0   186.0  

[285 rows x 15 columns]

Filtering cars which has weight lesser than 4000 from the region of Asia and Europe

In [ ]:
filtered_cars=origin_cars[~(origin_cars['Weight'] > 4000)] #this filters the cars with weight less than 4000
filtered_cars

Make                    Model   Type  Origin DriveTrain      MSRP  \
1    Acura           RSX Type S 2dr  Sedan    Asia      Front  $23,820    
2    Acura                  TSX 4dr  Sedan    Asia      Front  $26,990    
3    Acura                   TL 4dr  Sedan    Asia      Front  $33,195    
4    Acura               3.5 RL 4dr  Sedan    Asia      Front  $43,755    
5    Acura  3.5 RL w/Navigation 4dr  Sedan    Asia      Front  $46,100    
..     ...                      ...    ...     ...        ...       ...   
427  Volvo  C70 LPT convertible 2dr  Sedan  Europe      Front  $40,565    
428  Volvo  C70 HPT convertible 2dr  Sedan  Europe      Front  $42,565    
429  Volvo               S80 T6 4dr  Sedan  Europe      Front  $45,210    
430  Volvo                      V40  Wagon  Europe      Front  $26,135    
431  Volvo                     XC70  Wagon  Europe        All  $35,145    

      Invoice  EngineSize  Cylinders  Horsepower  MPG_City  MPG_Highway  \
1    $21,761          2.0        4.0       200.0      24.0         31.0   
2    $24,647          2.4        4.0       200.0      22.0         29.0   
3    $30,299          3.2        6.0       270.0      20.0         28.0   
4    $39,014          3.5        6.0       225.0      18.0         24.0   
5    $41,100          3.5        6.0       225.0      18.0         24.0   
..        ...         ...        ...         ...       ...          ...   
427  $38,203          2.4        5.0       197.0      21.0         28.0   
428  $40,083          2.3        5.0       242.0      20.0         26.0   
429  $42,573          2.9        6.0       268.0      19.0         26.0   
430  $24,641          1.9        4.0       170.0      22.0         29.0   
431  $33,112          2.5        5.0       208.0      20.0         27.0   

     Weight  Wheelbase  Length  
1    2778.0      101.0   172.0  
2    3230.0      105.0   183.0  
3    3575.0      108.0   186.0  
4    3880.0      115.0   197.0  
5    3893.0      115.0   197.0  
..      ...        ...     ...  
427  3450.0      105.0   186.0  
428  3450.0      105.0   186.0  
429  3653.0      110.0   190.0  
430  2822.0      101.0   180.0  
431  3823.0      109.0   186.0  

[231 rows x 15 columns]

## **Saving the dataset to Google drive Folder**

In [ ]:
filtered_cars.to_csv(r'/content/drive/My Drive/DA_project/Filtered_Cars_dataset', index=False)

Now we have stored the filtered car dataset which contains information about cars that weigh below 4000 and has origin from Asia and Europe